src/adtl/autoparser/dict_writer.py

```diff
@@ -0,0 +1,353 @@
+"""
+Infer a data dictionary from a dataset.
+"""
+
+from __future__ import annotations
+
+import argparse
+from pathlib import Path
+
+import google.generativeai as gemini
+import numpy as np
+import pandas as pd
+from openai import OpenAI
+
+from .gemini_calls import _get_definitions as _get_definitions_gemini
+from .openai_calls import _get_definitions as _get_definitions_openai
+from .util import DEFAULT_CONFIG, load_data_dict, read_config_schema, read_data
+
+
+class DictWriter:
+    """
+    Class for infering a data dictionary based on a dataset. Will not store the data,
+    only the created data dictionary.
+
+    Use `create_dict()` to create a data dictionary, as the function equivalent
+    of the command line `create-dict` script.
+
+    `generate_descriptions()` will use an LLM to generate descriptions for the
+    data dictionary, using only the column headers, NOT the data itself.
+
+    Parameters
+    ----------
+    config
+        The path to the configuration file to use if not using the default configuration
+    """
+
+    def __init__(
+        self,
+        config: Path | str | None = None,
+    ):
+        if isinstance(config, str):
+            config = Path(config)
+        self.config = read_config_schema(
+            config or Path(Path(__file__).parent, DEFAULT_CONFIG)
+        )
+
+    def _setup_llm(self, key: str, name: str):
+        """
+        Setup the LLM to use to generate descriptions.
+
+        Seperate from the __init__ method to allow for extra barrier between raw data &
+        LLM.
+
+        Parameters
+        ----------
+        key
+            API key
+        name
+            Name of the LLM to use (currently only OpenAI and Gemini are supported)
+        """
+        if key is None:
+            raise ValueError("API key required for generating descriptions")
+        else:
+            self.key = key
+
+        if name == "openai":  # pragma: no cover
+            self.client = OpenAI(api_key=key)
```

Comment:
which model does this use - good to have it explicitly mentioned

Reply:
This is specified in the 'openai_calls.py' file - perhaps when addressing your comment below about reworking the LLM calls with a class inheritance structure we can add in a defaulted option where the user can choose which model to use (i think InsightBoard does this, can follow a similar logic probably)